File: www/CiaoDE/ciao/library.development/proTcXl2.1/xlib.c
/* 
 * ProTcl - A Prolog Interface to the Tcl/Tk toolkit and to the
 *		Xlib library
 *
 *
 *	Copyright (c) 1995 Micha Meier
 *
 *******************************************************************/

#ifndef lint
static char     *SccsId   = "@(#)xlib.c	1.9        96/03/02";
#endif

#ifdef ECLIPSE
#include "tcl.h"
#include "tk.h"
#include "external.h"

#define Check_Xlib_Id(v, t, tkwin, gc)			\
	Error_If_Ref(t);				\
	if (IsStructure(t) && v.ptr->val.did == d_protcl_xlib_id)\
	{						\
		Check_Integer(v.ptr[1].tag)		\
		tkwin = (Tk_Window) v.ptr[1].val.nint;	\
		Check_Integer(v.ptr[2].tag)		\
		gc = (GC) v.ptr[2].val.nint;		\
	}						\
	else						\
	{						\
		Bip_Error(TYPE_ERROR);			\
	}

extern Tcl_Interp	*protcl_interp;
extern Tk_Window	protcl_mainWindow;
extern word32		d_protcl_xlib_id;

/*
 * xlib_init(+Path, -XlibID)
 *	Returns the XlibID of an existing Tk window.
 *	The XlibID is a structure xlib_id(Tk_Window, GC)
 */
p_xlib_init(vp, tp, vi, ti)
value	vp, vi;
type	tp, ti;
{
    Tk_Window			tkwin;
    GC				gc;
    XSetWindowAttributes	attr;
    char			*name;
    pword			*p;

    if (protcl_interp == 0) {
	Fail;
    }
    Get_Name(vp, tp, name)
    tkwin = Tk_NameToWindow(protcl_interp, name, protcl_mainWindow);
    if (tkwin == NULL) {
        Fail;
    }
    if (IsStructure(ti) && vi.ptr->val.did == d_protcl_xlib_id &&
		IsInteger(vi.ptr[2].tag))
	gc = (GC) vi.ptr[2].val.nint;
    else
	gc = XCreateGC(Tk_Display(tkwin), Tk_WindowId(tkwin), 0, 0);
    attr.backing_store = WhenMapped;
    Tk_ChangeWindowAttributes(tkwin, CWBackingStore, &attr);

    p = TG;
    TG += 3;
    Check_Gc;
    p[0].val.did = d_protcl_xlib_id;
    p[0].tag.all = TDICT;
    p[1].val.nint = (long) tkwin;
    p[1].tag.all = TINT;
    p[2].val.nint = (long) gc;
    p[2].tag.all = TINT;
    Return_Unify_Structure(vi, ti, p)
}

/*
 * xlib_id(+XlibID, -Display, -Window, -GC)
 *	Returns the Xlib items of a given Tk window
 */
p_xlib_id(vi, ti, vd, td, vw, tw, vg, tg)
value	vi, vd, vw, vg;
type	ti, td, tw, tg;
{
    Tk_Window	tkwin;
    GC		gc;
    Prepare_Requests;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    Request_Unify_Integer(vd, td, (long) Tk_Display(tkwin))
    Request_Unify_Integer(vw, tw, (long) Tk_WindowId(tkwin))
    Request_Unify_Integer(vg, tg, (long) gc)
    Succeed;
}

/*
 * xlib_tkwin(+XlibID, -TkWindow)
 *	Returns the Tk id of a given Tk window
 */
p_xlib_tkwin(vi, ti, vw, tw)
value	vi, vw;
type	ti, tw;
{
    Tk_Window	tkwin;
    GC		gc;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    Return_Unify_Integer(vw, tw, (long) tkwin)
}

/* Dummy change procedure for the image function */
void
chproc(cd, x, y, w, h, iw, ih)
ClientData cd;
int x;
int y;
int w;
int h;
int iw;
int ih;
{
}

/*
 * xlib_pixmap(+XlibID, +TkPixmap, -XlibPixmap)
 */
p_xlib_pixmap(vi, ti, vn, tn, vp, tp)
value           vi, vn, vp;
type            ti, tn, tp;
{
    char        *name;
    Pixmap      pixmap;
    Tk_Window   tkwin;
    GC          gc;
    Tk_Image    image;
    ClientData  cd;
    int         width;
    int         height;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    Get_Name(vn, tn, name);
    image = Tk_GetImage(protcl_interp, tkwin, name, chproc, cd);
    if (image == NULL) {
	Fail;
    }
    Tk_SizeOfImage(image, &width, &height);
    pixmap = Tk_GetPixmap(Tk_Display(tkwin), Tk_WindowId(tkwin),
        width, height, Tk_Depth(tkwin));
    Tk_RedrawImage(image, 0, 0, width, height, pixmap, 0, 0);
    Tk_FreeImage(image);
    Return_Unify_Integer(vp, tp, pixmap);
}

/*
 * xlib_color(+XlibID, +TkColor, -XlibColor)
 */
p_xlib_color(vi, ti, vc, tc, vp, tp)
value		vi, vc, vp;
type		ti, tc, tp;
{
    char	*color;
    Tk_Window	tkwin;
    Drawable	window;
    GC		gc;
    XColor	*xcolor;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    Get_Name(vc, tc, color);
    xcolor = Tk_GetColor(protcl_interp, tkwin, Tk_GetUid(color));
    if (xcolor == (XColor *) 0) {
    	Fail
    }
    Return_Unify_Integer(vp, tp, xcolor->pixel)
}

/*
 * xlib_font(+XlibID, +Font)
 */
p_xlib_font(vi, ti, vf, tf)
value		vi, vf;
type		ti, tf;
{
    char	*font;
    Tk_Window	tkwin;
    Drawable	window;
    GC		gc;
    XFontStruct	*xfont;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    Get_Name(vf, tf, font);
    xfont = Tk_GetFontStruct(protcl_interp, tkwin, Tk_GetUid(font));
    if (xfont == (XFontStruct *) 0) {
    	Fail
    }
    XSetFont(Tk_Display(tkwin), gc, xfont->fid);
    Succeed;
}

/*
 * xlib_query_pointer(+XlibID, -X, -Y)
 */
p_xlib_query_pointer(vi, ti, vx, tx, vy, ty)
value		vi, vx, vy;
type		ti, tx, ty;
{
    Tk_Window	tkwin;
    Drawable	window;
    GC		gc;
    Window	r1, r2;
    int		rx, ry;
    int		wx, wy;
    int		mask;
    int		res;
    Prepare_Requests;

    Check_Xlib_Id(vi, ti, tkwin, gc)
    res = XQueryPointer(Tk_Display(tkwin), Tk_WindowId(tkwin), &r1, &r2,
		&rx, &ry, &wx, &wy, &mask);
    if (res == False) {
    	Fail
    }
    Request_Unify_Integer(vx, tx, wx);
    Request_Unify_Integer(vy, ty, wy);
    Return_Unify;
}

/*
 * xlib_polygon(XlibID, DataList, OffX, OffY, RX, RY, Xpose, Border, Fill)
 *	draw a polygon with a given scale, offset, and colours
 */
p_xlib_polygon(iv, it, lv, lt, xv, xt, yv, yt, rxv, rxt, ryv, ryt, xpv, xpt, 
	bv, bt, fv, ft)
value	iv, lv, xv, yv, rxv, ryv, xpv, bv, fv;
type	it, lt, xt, yt, rxt, ryt, xpt, bt, ft;
{
    Tk_Window		tkwin;
    GC			gc;
    XPoint		*p, *start;
    pword		*l, *r;
    int			n = 0;
    register int	xpose = xpv.nint;
    unsigned long	color;

    Check_List(lt)
    Check_Xlib_Id(iv, it, tkwin, gc)
    Check_Integer(xt)
    Check_Integer(yt)
    Check_Float(rxt)
    Check_Float(ryt)
    Check_Integer(xpt)
    Check_Integer(bt)
    Check_Integer(ft)
    if (IsNil(lt)) {
	Succeed_
    }
    p = start = (XPoint *) Gbl_Tg;
    Gbl_Tg++;
    Check_Gc;
    l = lv.ptr;
    for (;;)
    {
	    r = l++;
	    Dereference_(r);
	    if (xpose)
		p->y = FloatVal(ryv, ryt) * (yv.nint + r->val.nint);
	    else
		p->x = FloatVal(rxv, rxt) * (xv.nint + r->val.nint);
	    Dereference_(l);
	    r = l->val.ptr;
	    l = r + 1;
	    Dereference_(r);
	    if (xpose)
		p->x = FloatVal(rxv, rxt) * (xv.nint + r->val.nint);
	    else
		p->y = FloatVal(ryv, ryt) * (yv.nint + r->val.nint);
	    n++;
	    if (p++ >= (XPoint *) Gbl_Tg) {
		    Gbl_Tg++;
		    Check_Gc;
	    }
	    Dereference_(l);
	    if (IsNil(l->tag))
		    break;
	    l = l->val.ptr;
    }

    if (fv.nint >= 0) {
	XSetForeground(Tk_Display(tkwin), gc, fv.nint);
	XFillPolygon(Tk_Display(tkwin), Tk_WindowId(tkwin), gc, start,
	    n, Nonconvex, 0);
    }
    if (bv.nint >= 0) {
	XSetForeground(Tk_Display(tkwin), gc, bv.nint);
	XDrawLines(Tk_Display(tkwin), Tk_WindowId(tkwin), gc, start, n, 0);
    }
    Gbl_Tg = (pword *) start;
    Succeed_
}

/* P is on the line drawn through (x,y) */
#define OnLine(px, py, x, y)	(y == py && px >= x)


/*
 *	inside(X, Y, Cover, List),
 *		Succeeds if the point [X, Y] is inside the polygon
 *		with coordinates in List = [X1, Y1, X2, Y2, ..., 
 */
p_inside(vx, tx, vy, ty, vc, tc, vl, tl)
value	vx, vy, vc, vl;
type	tx, ty, tc, tl;
{
	pword	*p;
	pword	*s;
	int	x, y, x1, y1, x2, y2, xl, yl;
	int	cx1, cx2, cy1, cy2;
	int	count;
	int	c1, c2;

	x = vx.nint;
	y = vy.nint;

	p = vc.ptr;
	s = p++;
	Dereference_(s);
	cx1 = s->val.nint;

	Dereference_(p);
	p = p->val.ptr;
	s = p++;
	Dereference_(s);
	cy1 = s->val.nint;

	Dereference_(p);
	p = p->val.ptr;
	s = p++;
	Dereference_(s);
	cx2 = s->val.nint;

	Dereference_(p);
	p = p->val.ptr;
	s = p++;
	Dereference_(s);
	cy2 = s->val.nint;

	/* check if in the square */
	if (x < cx1 || y < cy1 || x > cx2 || y > cy2) {
	    Fail_
	}
	p = vl.ptr;
	s = p++;
	Dereference_(s);
	xl = x2 = s->val.nint;
	Dereference_(p);

	p = p->val.ptr;
	s = p++;
	Dereference_(s);
	yl = y2 = s->val.nint;
	Dereference_(p);
	/* make sure that P2 is not on the line */
	if (OnLine(x2, y2, x, y)) {
	    if (x == x2) {
		Succeed;
	    }
	    xl = x2 = x2 - 1;
	}

	count = 0;
	while (IsList(p->tag))
	{
	    p = p->val.ptr;
	    s = p++;
	    Dereference_(s);
	    x1 = s->val.nint;
	    Dereference_(p);

	    p = p->val.ptr;
	    s = p++;
	    Dereference_(s);
	    y1 = s->val.nint;
	    Dereference_(p);

	    if (OnLine(x1, y1, x, y)) {
		if (x == x1) {
		    Succeed;
		}
		continue;
	    }
	    if (y1 >= y && y2 <= y || y1 <= y && y2 >= y) {
		c1 = (x2 - x1)*(y - y1);
		c2 = (y2 - y1)*(x - x1);
		if (y2 > y1 && c1 > c2 || y2 < y1 && c1 < c2)
		    count++;
	    }
	    x2 = x1;
	    y2 = y1;
	}
	if (!OnLine(x1, y1, x, y) && (y2 >= y && yl <= y || y2 <= y && yl >= y))
	{
	    c1 = (xl - x2)*(y - y2);
	    c2 = (yl - y2)*(x - x2);
	    if (yl > y2 && c1 > c2 || yl < y2 && c1 < c2)
		count++;
	}
	Succeed_If(count & 1);
}

/*
 * Dummy function to enforce linking of useful Xlib functions
 */
static
_dummy()
{
    Display	*d;
    Drawable	w;
    GC		gc;
    XGCValues	xv;

    XDrawImageString(d, w, gc, 0, 0, "", 0);
    XSetBackground(d, gc, 0);
    XWarpPointer(d, 0, 0, 0, 0, 0, 0, 0, 0);
    XChangeGC(d, gc, 0, &xv);
    XClearArea(d, w, 0, 0, 0, 0, 0);
    XSetLineAttributes(d, gc, 0, 0, 0, 0);
}

#endif /* ECLIPSE */

File: www/CiaoDE/ciao/library.development/proTcXl2.1/xlib.pl
%
% ECLiPSe version of the Prolog Tk interface
%
% Author: Micha Meier
% Date:   September 93
%

%
% sccsid("@(#)xlib.pl	1.5          96/03/02").
% sccscr("@(#)  Copyright 1996 Micha Meier ").
%

 %% Module et.al. interface deleted

 %% :- module_interface(xlib).
 %% 
 %% :- tool(xlib_define_macros/0).
 %% 
 %% :- export
 %% 	inside/4,
 %% 	tr_xlib/2,
 %% 	xlib_arc/7,
 %% 	xlib_background/2,
 %% 	xlib_change_gc/3,
 %% 	xlib_clear/5,
 %% 	xlib_color/3,
 %% 	xlib_define_macros/0,
 %% 	xlib_fill_arc/7,
 %% 	xlib_fill_rectangle/5,
 %% 	xlib_flush/1,
 %% 	xlib_font/2,
 %% 	xlib_foreground/2,
 %% 	xlib_function/2,
 %% 	xlib_id/4,
 %% 	xlib_image_string/4,
 %% 	xlib_init/2,
 %% 	xlib_line/5,
 %% 	xlib_line_attributes/5,
 %% 	xlib_pixmap/3,
 %% 	xlib_polygon/9,
 %% 	xlib_query_pointer/3,
 %% 	xlib_rectangle/5,
 %% 	xlib_string/4,
 %% 	xlib_tkwin/2,
 %% 	xlib_warp_pointer/3.

xlib_define_macros.		% dummy for the next query

:- xlib_define_macros.

 %% :- begin_module(xlib).
 %% 
 %% :- import define_macro_/4 from sepia_kernel.
 %% 
 %% :- [tr_xlib].
 %% 
 %% :- xlib_define_macros.
 %% 
 %% :- make_array(xlib_gcvalues(30), integer).
 %% 
 %% :- external(inside/4, p_inside).
 %% :- external(xlib_init/2, p_xlib_init).
 %% :- external(xlib_id/4, p_xlib_id).
 %% :- external(xlib_color/3, p_xlib_color).
 %% :- external(xlib_font/2, p_xlib_font).
 %% :- external(xlib_pixmap/3, p_xlib_pixmap).
 %% :- external(xlib_polygon/9, p_xlib_polygon).
 %% :- external(xlib_query_pointer/3, p_xlib_query_pointer).
 %% :- external(xlib_tkwin/2, p_xlib_tkwin).

foreign_file('xlib.o',
        [p_inside,
         p_xlib_init,
         p_xlib_id,
         p_xlib_color,
         p_xlib_font,
         p_xlib_pixmap,
         p_xlib_polygon,
         p_xlib_query_pointer,
         p_xlib_tkwin]).

 %% Arguments temptative.
 %% How should we treat the lists? 

foreign(p_inside, c, inside()).
foreign(p_xlib_init ,c , xlib_init()).
foreign(p_xlib_id ,c , xlib_id()).
foreign(p_xlib_color ,c , xlib_color()).
foreign(p_xlib_font ,c , xlib_font()).
foreign(p_xlib_pixmap ,c , xlib_pixmap()).
foreign(p_xlib_polygon ,c , xlib_polygon()).
foreign(p_xlib_query_pointer ,c , xlib_query_pointer()).
foreign(p_xlib_tkwin ,c , xlib_tkwin()).



%
% The most frequently used Xlib operations
%
xlib_flush(XID) :-
    xlib_id(XID, Display, _, _),
    call_c('XFlush'(Display), _).       % How is call_c defined?

xlib_foreground(XID, Color) :-
    xlib_color(XID, Color, XC),
    xlib_id(XID, Display, _, GC),
    call_c('XSetForeground'(Display, GC, XC), _).
    
xlib_background(XID, Color) :-
    xlib_color(XID, Color, XC),
    xlib_id(XID, Display, _, GC),
    call_c('XSetBackground'(Display, GC, XC), _).
    
xlib_function(XID, Function) :-
    xlib_id(XID, Display, _, GC),
    call_c('XSetFunction'(Display, GC, Function), _).
    
xlib_change_gc(XID, Name, Value) :-
    xlib_id(XID, Display, _, GC),
    Index0 is fix(ln(Name)/ln(2.0) + 0.1),
    % fix the arc_mode irregularity
    (Index0 < 10 ->
        Index = Index0
    ;
    Index0 = 22 ->
	Index = 10
    ;
        Index is Index0 + 1
    ),
    setval(xlib_gcvalues(Index), Value),
    call_c('XChangeGC'(Display, GC, Name, xlib_gcvalues/1), _).
    
xlib_image_string(XID, Thing, X, Y) :-
    xlib_id(XID, Display, Window, GC),
    id_length(Thing, String, Len),
    call_c('XDrawImageString'(Display, Window, GC, X, Y, String, Len), _).

xlib_string(XID, Thing, X, Y) :-
    xlib_id(XID, Display, Window, GC),
    id_length(Thing, String, Len),
    call_c('XDrawString'(Display, Window, GC, X, Y, String, Len), _).

xlib_arc(XID, X, Y, Width, Height, Angle1, Angle2) :-
    xlib_id(XID, Display, Win, GC),
    IA1 is floor(Angle1*64),
    IA2 is floor(Angle2*64),
    call_c('XDrawArc'(Display, Win, GC, X, Y, Width, Height, IA1, IA2), _).

xlib_fill_arc(XID, X, Y, Width, Height, Angle1, Angle2) :-
    xlib_id(XID, Display, Win, GC),
    IA1 is floor(Angle1*64),
    IA2 is floor(Angle2*64),
    call_c('XFillArc'(Display, Win, GC, X, Y, Width, Height, IA1, IA2), _).

xlib_clear(XID, X, Y, Width, Height) :-
    xlib_id(XID, Display, Win, _),
    call_c('XClearArea'(Display, Win, X, Y, Width, Height, 0), _).

xlib_line(XID, X1, Y1, X2, Y2) :-
    xlib_id(XID, Display, Win, GC),
    call_c('XDrawLine'(Display, Win, GC, X1, Y1, X2, Y2), _).

xlib_line_attributes(XID, Width, Style, Cap, Join) :-
    xlib_id(XID, Display, _, GC),
    call_c('XSetLineAttributes'(Display, GC, Width, Style, Cap, Join), _).

xlib_rectangle(XID, X, Y, Width, Height) :-
    xlib_id(XID, Display, Win, GC),
    call_c('XDrawRectangle'(Display, Win, GC, X, Y, Width, Height), _).

xlib_fill_rectangle(XID, X, Y, Width, Height) :-
    xlib_id(XID, Display, Win, GC),
    call_c('XFillRectangle'(Display, Win, GC, X, Y, Width, Height), _).

xlib_warp_pointer(XID, X, Y) :-
    xlib_id(XID, Display, Win, _),
    call_c('XWarpPointer'(Display, xNone, Win, 0, 0, 0, 0, X, Y), _).

xlib_stipple(XlibID, Bitmap) :-
    tk_id(XlibID, TkWin, Interp),
    xlib_id(XlibID, Display, _, GC),
    (Bitmap = "" ->
	call_c('XSetFillStyle'(Display, GC, xFillSolid))
    ;
	call_c('Tk_GetUid'(Bitmap), BitmapUid),
	call_c('Tk_GetBitmap'(Interp, TkWin, BitmapUid), Pixmap),
	call_c('XSetStipple'(Display, GC, Pixmap), _),
	call_c('XSetFillStyle'(Display, GC, xFillStippled), _)
    ).

id_length(A, A, L) :-
    atom(A),
    atom_length(A, L).
id_length(S, S, L) :-
    string(S),
    string_length(S, L).
id_length(N, S, L) :-
    number(N),
    number_string(N, S),
    id_length(S, _, L).
